File: 01-游戏开发/01-服务端/网络编程/协议设计/FlatBuffers零拷贝/python/main.py
# -*- coding: utf-8 -*-
"""FlatBuffers 线格式最小实现:构建 + 就地(零拷贝)读取。

复现 flatcc binary-format 文档的教学示例(FooBar):
    table FooBar { meal:byte=Banana(-1); density:long(deprecated);
                  say:string; height:short; }  file_identifier "NOOB"
数据 { meal:42(Orange), say:"hello", height:-8000 } 的官方参考字节。

布局规则(本教学构建器):table 内 uoffset 类字段在前(按字段 id 序)、
标量在后(按字段 id 序、自然对齐填充);vtable 放缓冲区末尾(flatcc 风格,
soffset 为负)。flatcc 文档明确:table 内字段顺序格式不规定,只要求对齐。
"""
import struct

BANANA, ORANGE = -1, 42


def ru16(b, o):
    return struct.unpack_from("<H", b, o)[0]


def ru32(b, o):
    return struct.unpack_from("<I", b, o)[0]


def ri32(b, o):
    return struct.unpack_from("<i", b, o)[0]


# ---------------- 构建器(flatcc 布局:block 顺序 header|table|string|vtable) ----------------

class Builder:
    """单 schema 教学构建器;同布局 vtable 在同一缓冲区内按字节串去重共享。"""

    def __init__(self, file_identifier=None):
        self.ident = file_identifier

    def build(self, fields, children):
        """fields: [(fid, kind, value, default)](kind: 'i16'|'i8',值==默认值则不存储);
        children: [(fid, bytes)](string 块,以 uoffset 引用)。
        返回 (bytes, table地址, vtable地址)。"""
        # 1. table 布局:soffset(4) 后先放 uoffset 槽(id 序),再放标量(id 序)
        body, uoff_slots, entries, off = [("soff", None)], [], {}, 4
        for fid, _ in children:
            uoff_slots.append((fid, off))
            entries[fid] = off
            body.append(("u32", None))
            off += 4
        for fid, kind, value, default in fields:
            if value == default:                 # 缺省值不占存储,vtable 条目 0
                continue
            size = 2 if kind == "i16" else 1
            if off % size:
                body.append(("pad", size - off % size))
                off += size - off % size
            entries[fid] = off
            body.append(("i16" if kind == "i16" else "i8", value))
            off += size
        table_size = off
        # 2. 组装 block:header | table | children | 填充 | vtable
        blocks = []
        table_pos = 4 + (4 if self.ident else 0)
        for kind, value in body:
            if kind == "pad":
                blocks.append(b"\x00" * value)
            elif kind in ("soff", "u32"):
                blocks.append(b"\x00\x00\x00\x00")     # 占位,稍后回填
            elif kind == "i16":
                blocks.append(struct.pack("<h", value))
            else:
                blocks.append(struct.pack("<b", value))
        end = table_pos + sum(len(x) for x in blocks)
        child_pos = {}
        for fid, data in children:                      # string 块:长度+内容+\0
            while end % 4:
                blocks.append(b"\x00")
                end += 1
            child_pos[fid] = end
            blocks.append(struct.pack("<I", len(data)) + data + b"\x00")
            end += 4 + len(data) + 1
        while end % 4:
            blocks.append(b"\x00")
            end += 1
        # 3. vtable:vtable_size, table_size, 各字段偏移(0=缺失)
        max_fid = max(list(entries) + [0])
        vt_list = [2 * (max_fid + 3), table_size] + [
            entries.get(fid, 0) for fid in range(max_fid + 1)]
        vt_bytes = struct.pack("<%dH" % len(vt_list), *vt_list)
        vtable_pos = end
        blocks.append(vt_bytes)
        # 4. 拼接并回填:soffset(负向)、uoffset(正向)
        buf = bytearray(struct.pack("<I", table_pos))
        if self.ident:
            buf += self.ident
        buf += b"".join(blocks)
        struct.pack_into("<i", buf, table_pos, table_pos - vtable_pos)
        for fid, off in uoff_slots:
            struct.pack_into("<I", buf, table_pos + off,
                             child_pos[fid] - (table_pos + off))
        return bytes(buf), table_pos, vtable_pos


# ---------------- 读取器(就地访问,零拷贝) ----------------

def field_offset(buf, t, fid):
    """vtable 查字段:越界(旧数据无该字段)或条目 0 → 返回 0 表示缺失。"""
    vt = t - ri32(buf, t)                                # soffset 是『减去』
    vtsize = ru16(buf, vt)
    slot = 4 + fid * 2      # 跳过 vtable 长度/表长度两个头条目
    if slot >= vtsize:
        return 0
    return ru16(buf, vt + slot)


def get_scalar(buf, t, fid, fmt, default):
    off = field_offset(buf, t, fid)
    if off == 0:
        return default, False
    return struct.unpack_from(fmt, buf, t + off)[0], True


def get_string_view(buf, t, fid, default=None):
    """返回 memoryview 视图(零拷贝):经 view 改写直接落到原缓冲区。"""
    off = field_offset(buf, t, fid)
    if off == 0:
        return default
    s = t + off + ru32(buf, t + off)                     # uoffset 加在自身存储地址上
    n = ru32(buf, s)
    assert s + 4 + n <= len(buf), "string 越界"
    return memoryview(buf)[s + 4: s + 4 + n]


# ---------------- 断言 ----------------

def check(label, cond, detail=""):
    assert cond, "%s %s" % (label, detail)
    print("[ok] %s" % label)


def main():
    # ---- 断言 1:逐字节复现 flatcc 参考缓冲区 ----
    buf, t, vt = Builder(b"NOOB").build(
        fields=[(0, "i8", ORANGE, BANANA), (3, "i16", -8000, 0)],
        children=[(2, b"hello")],
    )
    expected = (struct.pack("<I", 8) + b"NOOB"
                + struct.pack("<i", -24) + struct.pack("<I", 8)
                + struct.pack("<b", ORANGE) + b"\x00" + struct.pack("<h", -8000)
                + struct.pack("<I", 5) + b"hello\x00" + b"\x00\x00"
                + struct.pack("<6H", 12, 12, 8, 0, 4, 10))
    check("builder 复现 flatcc 参考字节", buf == expected, "got %s" % buf.hex())

    # ---- 断言 2:就地读取 ----
    check("root uoffset -> table @8", ru32(buf, 0) == 8)
    check("soffset(负) -> vtable @0x20", vt == 0x20 and t - ri32(buf, t) == vt)
    meal, present = get_scalar(buf, t, 0, "<b", BANANA)
    check("meal=42(Orange)", meal == ORANGE and present)
    height, _ = get_scalar(buf, t, 3, "<h", 0)
    check("height=-8000(int16 小端)", height == -8000)
    check("say=='hello'", bytes(get_string_view(buf, t, 2)) == b"hello")
    check("deprecated density 条目为 0(缺失)", field_offset(buf, t, 1) == 0)

    # ---- 断言 3:零拷贝 —— view 直接落在原缓冲区上 ----
    mut = bytearray(buf)
    get_string_view(mut, t, 2)[0] = ord("j")
    check("零拷贝:经 view 改写字节直落缓冲区", mut[0x18] == ord("j"))

    # ---- 断言 4:缺省值不存储 + 前向兼容读取 ----
    buf2, t2, _ = Builder(b"NOOB").build(
        fields=[(0, "i8", BANANA, BANANA), (3, "i16", -8000, 0)],
        children=[(2, b"hi")])
    check("值==默认值时 vtable 条目置 0", field_offset(buf2, t2, 0) == 0)
    meal2, present2 = get_scalar(buf2, t2, 0, "<b", BANANA)
    check("缺省读取返回默认 -1 且不占存储", meal2 == BANANA and not present2)
    level, present4 = get_scalar(buf, t, 4, "<i", 7)     # 新 schema 的 fid=4
    check("新代码读旧数据:越界字段返回默认", level == 7 and not present4)

    # ---- 断言 5:同布局两表共享同一 vtable(手工拼装单缓冲区) ----
    # header(4) | tableA(8) | tableB(8) | vtable(8):字段 f0:i16@4, f1:i8@6
    vt_shared = struct.pack("<4H", 8, 8, 4, 6)
    table = lambda so: struct.pack("<i", so) + struct.pack("<h", 11) \
        + struct.pack("<b", 2) + b"\x00"
    buf3 = struct.pack("<I", 4) + table(-16) + table(-8) + vt_shared
    tA, tB = 4, 12
    check("两表 soffset 均解析到同一 vtable @20",
          tA - ri32(buf3, tA) == 20 and tB - ri32(buf3, tB) == 20)
    check("vtable 在缓冲区中仅出现一份", buf3.count(vt_shared) == 1)

    # ---- 断言 6:vector(长度=元素个数)读取 + 截断防护 ----
    # header|table(soffset+uoffset)|vtable|填充|vector: f0 -> [10,20,30]
    bufv = (struct.pack("<I", 4) + struct.pack("<i", -8)
            + struct.pack("<I", 12) + struct.pack("<3H", 6, 8, 4)
            + b"\x00\x00" + struct.pack("<I", 3)
            + struct.pack("<3h", 10, 20, 30))
    vec = 4 + field_offset(bufv, 4, 0) + ru32(bufv, 4 + 4)
    check("vector 长度字段=元素个数 3", ru32(bufv, vec) == 3)
    items = struct.unpack_from("<3h", bufv, vec + 4)
    check("vector 内容 [10,20,30]", items == (10, 20, 30))
    try:
        get_string_view(buf[:6], 8, 2)                  # 小于最小长度 8 字节
        raised = False
    except struct.error:
        raised = True
    check("截断缓冲区读取触发越界异常", raised)

    print("\n全部断言通过")


if __name__ == "__main__":
    main()

File: 01-游戏开发/01-服务端/网络编程/协议设计/FlatBuffers零拷贝/go/main.go
// FlatBuffers 线格式最小实现(Go 版):构建 + 就地(零拷贝)读取。
// 复现 flatcc binary-format 文档教学示例(FooBar):数据
// { meal:42(Orange), say:"hello", height:-8000 } 的官方参考字节。
// 布局规则与 python/main.py 相同:uoffset 字段在前、标量按 id 序,
// vtable 放缓冲区末尾(flatcc 风格,soffset 为负)。
package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"os"
)

const (
	banana int8 = -1
	orange int8 = 42
)

func ru16(b []byte, o int) uint16 { return binary.LittleEndian.Uint16(b[o:]) }
func ru32(b []byte, o int) uint32 { return binary.LittleEndian.Uint32(b[o:]) }
func ri32(b []byte, o int) int32  { return int32(ru32(b, o)) }
func ri16(b []byte, o int) int16  { return int16(ru16(b, o)) }

// buildFooBar:单 schema 构建器。fields: meal(i8, id0, 默认 banana),
// height(i16, id3, 默认 0);children: say(string, id2)。
// 值等于默认值的字段不存储(vtable 条目 0),后续标量前移补位。
func buildFooBar(ident []byte, meal int8, height int16, say []byte) []byte {
	// 1. table 布局:soffset(4) | say uoffset(4) | meal(1)+填充 | height(2)
	mealStored, heightStored := meal != banana, height != 0
	off := 4 + 4 // soffset + say uoffset
	mealOff, heightOff := 0, 0
	if mealStored {
		mealOff = off
		off++
	}
	if heightStored {
		if off%2 != 0 {
			off++ // 对齐填充
		}
		heightOff = off
		off += 2
	}
	tableSize := off
	// 2. block 地址规划:header | table | string | 填充 | vtable
	tablePos := 4 + len(ident)
	end := tablePos + tableSize
	for end%4 != 0 {
		end++
	}
	strPos := end
	end += 4 + len(say) + 1
	for end%4 != 0 {
		end++
	}
	vtPos := end
	vtSize := 12 // vtable_size, table_size, meal, density, say, height 共 6 条目
	// 3. 拼装
	buf := make([]byte, vtPos+vtSize)
	binary.LittleEndian.PutUint32(buf[0:], uint32(tablePos)) // root uoffset
	copy(buf[4:], ident)
	binary.LittleEndian.PutUint32(buf[tablePos:], uint32(tablePos-vtPos)) // soffset(负)
	binary.LittleEndian.PutUint32(buf[tablePos+4:], uint32(strPos-(tablePos+4)))
	if mealStored {
		buf[tablePos+mealOff] = byte(meal)
	}
	if heightStored {
		binary.LittleEndian.PutUint16(buf[tablePos+heightOff:], uint16(height))
	}
	binary.LittleEndian.PutUint32(buf[strPos:], uint32(len(say)))
	copy(buf[strPos+4:], say)
	binary.LittleEndian.PutUint16(buf[vtPos:], uint16(vtSize))
	binary.LittleEndian.PutUint16(buf[vtPos+2:], uint16(tableSize))
	binary.LittleEndian.PutUint16(buf[vtPos+4:], uint16(mealOff))
	binary.LittleEndian.PutUint16(buf[vtPos+6:], 0) // density: deprecated/缺失
	binary.LittleEndian.PutUint16(buf[vtPos+8:], 4)
	binary.LittleEndian.PutUint16(buf[vtPos+10:], uint16(heightOff))
	return buf
}

// fieldOffset:vtable 查字段。越界(旧数据无该字段)或条目 0 → 返回 0 表示缺失。
func fieldOffset(buf []byte, t, fid int) int {
	vt := t - int(ri32(buf, t)) // soffset 是『减去』
	vtsize := int(ru16(buf, vt))
	slot := 4 + fid*2 // 跳过 vtable 长度/表长度两个头条目
	if slot >= vtsize {
		return 0
	}
	return int(ru16(buf, vt+slot))
}

func getScalarI8(buf []byte, t, fid int, def int8) (int8, bool) {
	off := fieldOffset(buf, t, fid)
	if off == 0 {
		return def, false
	}
	return int8(buf[t+off]), true
}

func getScalarI16(buf []byte, t, fid int, def int16) (int16, bool) {
	off := fieldOffset(buf, t, fid)
	if off == 0 {
		return def, false
	}
	return ri16(buf, t+off), true
}

// getString:返回共享底层数组的切片(零拷贝:改 buf 即改字符串内容)。
func getString(buf []byte, t, fid int) []byte {
	off := fieldOffset(buf, t, fid)
	if off == 0 {
		return nil
	}
	s := t + off + int(ru32(buf, t+off)) // uoffset 加在自身存储地址上
	n := int(ru32(buf, s))
	return buf[s+4 : s+4+n]
}

var failures int

func check(label string, cond bool) {
	if !cond {
		failures++
		fmt.Printf("[FAIL] %s\n", label)
		return
	}
	fmt.Printf("[ok] %s\n", label)
}

func main() {
	// ---- 断言 1:逐字节复现 flatcc 参考缓冲区 ----
	buf := buildFooBar([]byte("NOOB"), orange, -8000, []byte("hello"))
	expected := append([]byte{},
		0x08, 0x00, 0x00, 0x00, 'N', 'O', 'O', 'B',
		0xe8, 0xff, 0xff, 0xff, 0x08, 0x00, 0x00, 0x00,
		0x2a, 0x00, 0xc0, 0xe0, 0x05, 0x00, 0x00, 0x00,
		'h', 'e', 'l', 'l', 'o', 0x00, 0x00, 0x00,
		0x0c, 0x00, 0x0c, 0x00, 0x08, 0x00, 0x00, 0x00,
		0x04, 0x00, 0x0a, 0x00)
	check("builder 复现 flatcc 参考字节", bytes.Equal(buf, expected))

	// ---- 断言 2:就地读取 ----
	t := int(ru32(buf, 0))
	check("root uoffset -> table @8", t == 8)
	check("soffset(负) -> vtable @0x20", t-int(ri32(buf, t)) == 0x20)
	meal, present := getScalarI8(buf, t, 0, banana)
	check("meal=42(Orange)", meal == orange && present)
	height, _ := getScalarI16(buf, t, 3, 0)
	check("height=-8000(int16 小端)", height == -8000)
	check("say=='hello'", string(getString(buf, t, 2)) == "hello")
	check("deprecated density 条目为 0(缺失)", fieldOffset(buf, t, 1) == 0)

	// ---- 断言 3:零拷贝 —— 切片直接落在原缓冲区上 ----
	view := getString(buf, t, 2)
	view[0] = 'j'
	check("零拷贝:经切片改写字节直落缓冲区", buf[0x18] == 'j')
	view[0] = 'h'

	// ---- 断言 4:缺省值不存储 + 前向兼容读取 ----
	buf2 := buildFooBar([]byte("NOOB"), banana, -8000, []byte("hi"))
	check("值==默认值时 vtable 条目置 0", fieldOffset(buf2, int(ru32(buf2, 0)), 0) == 0)
	meal2, present2 := getScalarI8(buf2, int(ru32(buf2, 0)), 0, banana)
	check("缺省读取返回默认 -1 且不占存储", meal2 == banana && !present2)
	_, present4 := getScalarI16(buf, t, 4, 7) // 新 schema 的 fid=4
	check("新代码读旧数据:越界字段返回默认", !present4)

	// ---- 断言 5:同布局两表共享同一 vtable(手工拼装单缓冲区) ----
	// header(4) | tableA(8) | tableB(8) | vtable(8):字段 f0:i16@4, f1:i8@6
	vtShared := []byte{0x08, 0x00, 0x08, 0x00, 0x04, 0x00, 0x06, 0x00}
	mkTable := func(so int32) []byte {
		b := make([]byte, 8)
		binary.LittleEndian.PutUint32(b[0:], uint32(so))
		binary.LittleEndian.PutUint16(b[4:], 11)
		b[6] = 2
		return b
	}
	buf3 := append(append(append([]byte{0x04, 0x00, 0x00, 0x00},
		mkTable(-16)...), mkTable(-8)...), vtShared...)
	tA, tB := 4, 12
	check("两表 soffset 均解析到同一 vtable @20",
		tA-int(ri32(buf3, tA)) == 20 && tB-int(ri32(buf3, tB)) == 20)
	check("vtable 在缓冲区中仅出现一份", bytes.Count(buf3, vtShared) == 1)

	// ---- 断言 6:vector(长度=元素个数)读取 ----
	// header|table(soffset+uoffset)|vtable|填充|vector: f0 -> [10,20,30]
	bufv := []byte{
		0x04, 0x00, 0x00, 0x00, // root uoffset
		0xf8, 0xff, 0xff, 0xff, // soffset = -8 -> vtable @12
		0x0c, 0x00, 0x00, 0x00, // uoffset = 12 -> vector 长度字段 @20
		0x06, 0x00, 0x08, 0x00, 0x04, 0x00, // vtable: size=6, table=8, f0=+4
		0x00, 0x00, // 对齐填充
		0x03, 0x00, 0x00, 0x00, // vector 元素个数 = 3
		0x0a, 0x00, 0x14, 0x00, 0x1e, 0x00, // [10, 20, 30]
	}
	vec := 4 + fieldOffset(bufv, 4, 0) + int(ru32(bufv, 8))
	check("vector 长度字段=元素个数 3", ru32(bufv, vec) == 3)
	items := []int16{ri16(bufv, vec+4), ri16(bufv, vec+6), ri16(bufv, vec+8)}
	check("vector 内容 [10,20,30]", items[0] == 10 && items[1] == 20 && items[2] == 30)

	if failures > 0 {
		fmt.Printf("\n%d 项断言失败\n", failures)
		os.Exit(1)
	}
	fmt.Println("\n全部断言通过")
}
